File: src/main/java/fortress/transformers/NnfTransformer.java
package fortress.transformers;

import fortress.tfol.*;

public class NnfTransformer implements TheoryTransformer {
    @Override
    public Theory apply(Theory theory) {
        Signature sig = theory.getSignature();
        Theory result = Theory.mkTheoryWithSignature(sig);
        
        for(Term axiom : theory.getAxioms()) {
            Term axiomPrime = axiom.nnf(sig);
            result = result.withAxiom(axiomPrime);
        }
        
        return result;
    }
}

File: src/main/scala/fortress/transformers/NnfTransformer.scala
package fortress.transformers

import fortress.msfol._
import fortress.operations.TermOps._

/** Changes each axiom of the theory into negation normal form.
  * All other aspects of the theory remain unchanged.
  * The resulting theory is equivalent to the first (that is, corresponding formulas
  * are equivalent. */
class NnfTransformer extends TheoryTransformer {
    override def apply(theory: Theory): Theory = {
        val newAxioms = theory.axioms map (_.nnf)
        theory.withoutAxioms
            .withAxioms(newAxioms)
    }
    
    override def name: String = "Negation Normal Form Transformer"
}
